### PR TITLE
feat: add `plain` format type for customDatasources

### PR DESCRIPTION
**Pull request by arnabrahman**
## Changes

This PR has added a new option for format: `plain` for `customDatasources`. If the plain format is used then Renovate calls the endpoint with the ContentType header value text/plain.

The returned body is split per line and each line is then converted into a JSON object with `"version": "<trimmed line content>"`

Body of the response:
```
1.0.0
2.0.0
3.0.0
```
Should be converted to:
```
{
  "releases": [
    {
      "version": "1.0.0"
    },
    {
      "version": "2.0.0"
    },
    {
      "version": "3.0.0"
    }
  ]
}
```
## Context
This is well described in issue 
Closes https://github.com/renovatebot/renovate/issues/23371

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/custom/index.spec.ts
Comment (secustor) [inline lib/modules/datasource/custom/index.spec.ts:1]: Add a test case for when the endpoint only returns a single version, as well if the response body is not what is expected.

**Review by secustor — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Please test on a real repo and add documentation for the `plain` format in the datasource README.

**Comment by arnabrahman** [comment]
I followed this(https://github.com/renovatebot/renovate/blob/main/docs/development/local-development.md#run-against-a-real-repo) to test on a real repo, I think this is what you meant by testing on real repo. 

And about the documentation, do you think this is enough or should I add the description you mentioned in the issue? Your suggestion is appreciated. 

Sorry Sebastian for pinging you here, just want to be sure. @secustor

File: lib/modules/datasource/custom/index.ts [hunk 1]
```diff
@@ -64,4 +68,24 @@ export class CustomDatasource extends Datasource {
       return null;
     }
   }
+
+  private async fetchPlainFormat(url: string): Promise<unknown> {
+    const response = await this.http.get(url, {
+      headers: {
+        'Content-Type': 'text/plain',
```
Comment (secustor) [inline lib/modules/datasource/custom/index.ts:75]:
```suggestion
        'Accept': 'text/plain',
```
The client sends the `Accept` header to get the same back as `Content-Type`

**Review by secustor — CHANGES_REQUESTED** [review]
Yes, pleas add the description of the issue to the documentation.

It is not clear what is fed to the jsonata engine without it.

File: lib/modules/datasource/custom/index.ts [hunk 2]
```diff
@@ -64,4 +68,24 @@ export class CustomDatasource extends Datasource {
...
+      },
+    });
+    const contentType = response.headers['content-type'];
+    if (!contentType?.startsWith('text/')) {
+      return null;
+    }
+    const versions = response.body.split('\n').map((version) => {
```
Comment (viceice) [inline lib/modules/datasource/custom/index.ts:82]: use newline regex from regex utils

File: lib/modules/datasource/custom/readme.md [hunk 3]
```diff
@@ -78,6 +78,38 @@ All available options:
 }
 ```
 
+`Plain Format:`
```
Comment (secustor) [inline lib/modules/datasource/custom/readme.md:81]:
```suggestion
### Formats

#### JSON
If `json` is used processing works as described above.    
The returned body will be directly interpreted as JSON and forwarded to the transformation rules.

#### Plain
```

**Review by secustor — DISMISSED** [review]

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -663,7 +663,7 @@ As this is a template it can be dynamically set. E.g. add the `packageName` as p
 ### format
 
 Defines which format the API is returning.
-Only `json` is supported, but more are planned for future.
+At this moment, `json` & `plain` are supported, but more formats are planned for the future
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:666]:
```suggestion
The API returns the `json` or `plain` format types.
```

Changes:

- Replace `&` with plain English `and`
- Remove "promise" to support more format types in future

I think it's best to say what we support _now_, and leave any future things be.

Question: how do you set the `format` you get from the API? I don't see a code example here. Or is this something that's obvious to "real programmers"? 😄

File: lib/modules/datasource/custom/index.spec.ts [hunk 5]
```diff
@@ -84,6 +84,120 @@ describe('modules/datasource/custom/index', () => {
       expect(result).toEqual(expected);
     });
 
+    it('return releases for plain text api directly exposing in renovate format', async () => {
```
Comment (HonkingGoose) [inline lib/modules/datasource/custom/index.spec.ts:87]:
```suggestion
    it('return releases for plain text API directly exposing in Renovate format', async () => {
```

File: lib/modules/datasource/custom/index.spec.ts [hunk 6]
```diff
@@ -84,6 +84,120 @@ describe('modules/datasource/custom/index', () => {
...
+          foo: {
+            defaultRegistryUrlTemplate: 'https://example.com/v1',
+            format: 'plain',
+          },
+        },
+      });
+      expect(result).toEqual(expected);
+    });
+
+    it('return releases for plain text api and trim the content', async () => {
```
Comment (HonkingGoose) [inline lib/modules/datasource/custom/index.spec.ts:120]:
```suggestion
    it('return releases for plain text API and trim the content', async () => {
```

File: lib/modules/datasource/custom/index.spec.ts [hunk 7]
```diff
@@ -84,6 +84,120 @@ describe('modules/datasource/custom/index', () => {
...
+          foo: {
+            defaultRegistryUrlTemplate: 'https://example.com/v1',
+            format: 'plain',
+          },
+        },
+      });
+      expect(result).toEqual(expected);
+    });
+
+    it('return releases for plain text api when only returns a single version', async () => {
```
Comment (HonkingGoose) [inline lib/modules/datasource/custom/index.spec.ts:153]:
```suggestion
    it('return releases for plain text API when only returns a single version', async () => {
```

File: lib/modules/datasource/custom/index.spec.ts [hunk 8]
```diff
@@ -84,6 +84,120 @@ describe('modules/datasource/custom/index', () => {
...
+          foo: {
+            defaultRegistryUrlTemplate: 'https://example.com/v1',
+            format: 'plain',
+          },
+        },
+      });
+      expect(result).toEqual(expected);
+    });
+
+    it('return null for plain text api if the body is not what is expected', async () => {
```
Comment (HonkingGoose) [inline lib/modules/datasource/custom/index.spec.ts:177]:
```suggestion
    it('return null for plain text API if the body is not what is expected', async () => {
```

File: lib/modules/datasource/custom/readme.md [hunk 9]
```diff
@@ -78,6 +78,44 @@ All available options:
 }
 ```
 
+### Formats
+
+#### JSON
```
Comment (HonkingGoose) [inline lib/modules/datasource/custom/readme.md:83]:
```suggestion
#### JSON

```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md (hunk 4 above)
Comment (arnabrahman) [reply docs/usage/configuration-options.md]: @HonkingGoose I think you can get the examples from here https://github.com/renovatebot/renovate/pull/23563/files#diff-be0a39f342380493f17c380bbd694f1f05d18bdd8d7cecf174369084952102deR646

**Review by HonkingGoose — APPROVED** [review]
I'm happy enough with the style/grammar of the docs. I'll let the maintainers review for technical accuracy.

File: docs/usage/configuration-options.md [hunk 10]
```diff
@@ -663,7 +663,7 @@ As this is a template it can be dynamically set. E.g. add the `packageName` as p
 ### format
 
 Defines which format the API is returning.
-Only `json` is supported, but more are planned for future.
+The API returns the `json` or `plain` format types.
```
Comment (secustor) [inline docs/usage/configuration-options.md:666]:
```suggestion
Currently `json` or `plain` are supported, see the `custom` [datasource documentation](/modules/datasource/custom/) for more information.
```

**Review by secustor — APPROVED** [review]

**Pull request merged**